popdat/src/make_person.rs

```diff
@@ -1,158 +1,221 @@
-            "social_centre",
-            "studio",
-            "theatre",
-            "bar",
-            "bbq",
-            "bicycle_rental",
-            "boat_rental",
-            "boat_sharing",
-            "dive_centre",
-            "internet_cafe",
-        ],
-        Activity::Errands => vec![
-            "marketplace",
-            "post_box",
-            "photo_booth",
-            "recycling",
-            "townhall",
-        ],
-        Activity::Financial => vec!["bank", "atm", "bureau_de_change"],
-        Activity::Healthcare => vec![
-            "baby_hatch",
-            "clinic",
-            "dentist",
-            "doctors",
-            "hospital",
-            "nursing_home",
-            "pharmacy",
-            "social_facility",
-            "veterinary",
-            "childcare",
-        ],
-        Activity::Work => vec!["bank", "clinic"],
-        // TODO Fill this out. amenity_type in map_gui/src/tools/mod.rs might be helpful. It might
-        // also be helpful to edit the list of possible activities in lib.rs too.
-        _ => vec![],
-    }
-    .into_iter()
-    .collect();
-
-    // Find all buildings with a matching amenity
-    let mut candidates: Vec<BuildingID> = Vec::new();
-    for b in map.all_buildings() {
-        for amenity in &b.amenities {
-            if categories.contains(amenity.amenity_type.as_str()) {
-                candidates.push(b.id);
+        // TODO Others to fill out. amenity_type in map_gui/src/tools/mod.rs might be helpful. It
+        // might also be helpful to edit the list of possible activities in lib.rs too.
+
+        // Find all buildings with a matching amenity
+        let mut candidates: HashMap<Activity, Vec<BuildingID>> = HashMap::new();
+        for b in map.all_buildings() {
+            for (activity, categories) in &categories {
+                for amenity in &b.amenities {
+                    if categories.contains(&amenity.amenity_type.as_str()) {
+                        candidates
+                            .entry(*activity)
+                            .and_modify(|v| v.push(b.id))
+                            .or_insert(vec![b.id]);
+                    }
+                }
             }
         }
+        candidates
+    }
+
+    fn find_building_for_activity(
+        &self,
+        activity: Activity,
+        _start: TripEndpoint,
+        _map: &Map,
+        rng: &mut XorShiftRng,
+    ) -> Option<BuildingID> {
+        // TODO If there are several choices of building that satisfy an activity, which one will
+        // someone choose? One simple approach could just calculate the difficulty of going from the
+        // previous location (starting from home) to that place, using some mode of travel. Then
+        // either pick the closest choice, or even better, randomize, but weight based on
+        // the cost of getting there. map.pathfind() may be helpful.
+
+        // For now, just pick a random one
```

Comment:
behavioral change: I switched this to random rather than first.

Reply:
SGTM